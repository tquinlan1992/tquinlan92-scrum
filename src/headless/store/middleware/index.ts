// import saveToLocalStorageState from './saveToLocalStorageState';
import thunk from 'redux-thunk';
import { routerÂ } from './router';
import logger from 'redux-logger';

export const middleware = [
    // saveToLocalStorageState,
    router,
    thunk,
    logger
];
