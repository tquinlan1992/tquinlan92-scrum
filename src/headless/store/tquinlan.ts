import { Dictionary, mapValues, omit, merge } from "lodash";
import { Reducer, AnyAction } from "redux";
import { makeSimpleReducer, ActionCreator } from "tquinlan92-typescript-redux-utils";

export function getActions<T extends { [key: string]: {actions?: Dictionary<any>} }>(creators: T): { [P in keyof T]: T[P]['actions'] } {
    return mapValues(creators, "actions") as { [P in keyof T]: T[P]['actions'] };
} 

export function makeNestedSimpleReducerSimpleActions<AppState>(state: any) {
    const actionsReducers = mapValues(state, (value, key) => {
        return makeSimpleReducer(key, omit(value, 'actions'));
    });//{ [P in keyof T]: T[P]['actionCreator'] 
    const reducers = mapValues(actionsReducers, 'reducer');
    const actions = getActions(actionsReducers);

    return {
        actions,
        reducers
    } as {
        reducers: {
            [P in keyof AppState]: Reducer<AppState[P], AnyAction>
        };
        actions: {
            [P in keyof AppState]: {
                [A in keyof AppState[P]]: ActionCreator<AppState[P][A]>
            } & {
                reset: ActionCreator<null>;
                setAll: ActionCreator<AppState[P]>;
                set: ActionCreator<Partial<AppState[P]>>;
            } & {
                [A in keyof AppState[P]]: AppState[P][A]
            }
        };
    };
}

export function makeNestedSimpleStore<State, ThunkActions>(state: State, thunkActions?: ThunkActions) {
    const { actions: actions1, reducers: reducers2 } = makeNestedSimpleReducerSimpleActions<State>(state);
    const actionsWithThunks = merge(actions1, thunkActions);
    return {
        actions: actionsWithThunks,
        reducers: reducers2
    }
}