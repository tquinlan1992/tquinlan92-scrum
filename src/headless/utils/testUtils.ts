export function mockPouchDB() {
    jest.mock('../database/core/pouch', () => {
        return {};
    });
}
