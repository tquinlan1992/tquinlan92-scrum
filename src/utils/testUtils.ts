export function mockPouchDB() {
    jest.mock('../headless/database/pouch', () => {
        return {};
    });
}

export function mockPouch(dbMethods: object) {
    jest.mock('../headless/database/pouch', () => {
        return {

        }
    })
    const mockPouchRequire = require('../headless/database/pouch');
    const mockGetRemoteDb = jest.fn(() => {
        return new Promise(resolve => {
            resolve(dbMethods);
        })
    });
    mockPouchRequire.getRemoteDb = mockGetRemoteDb;
    return mockPouchRequire;
}

export function mockShortId() {
    jest.mock('shortid', () => {
        let count = 0;
        return () => {
            count = count + 1;
            return String(count);
        }
    })
}

