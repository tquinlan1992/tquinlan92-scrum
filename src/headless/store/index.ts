import { createStore, applyMiddleware, AnyAction, Store, combineReducers } from "redux";
import { middleware } from './middleware';
import { composeWithDevTools } from 'redux-devtools-extension';
import { routerReducer, RouterState } from "react-router-redux";
import { TicketListState } from '@components/TicketList/redux';
import { AddTicketState, addTicketDialogReducer } from '@components/AddTicketDialog/redux';
import { LoadingState, loadingReducer } from '@components/Loading/redux';
import { AppState } from '../store';
import { ticketListReducer } from '../../components/TicketList/redux';
import { ThunkAction } from "redux-thunk";
import { CloseSprintDialogState, closeSprintDialogReducer } from '@components/CloseSprintDialog/redux';


const appReducer = combineReducers<AppState>({
    routing: routerReducer,
    addTicket: addTicketDialogReducer,
    ticketList: ticketListReducer,
    loading: loadingReducer,
    closeSprintDialog: closeSprintDialogReducer
});

declare module 'redux' {
    export type GenericStoreEnhancer = any;
}

export type AppThunkAction = ThunkAction<void, AppState, void, AnyAction>;

export interface Doc {
    edit: string;
}

export interface VoiceToCode {
    text: string;
}

export interface AddTicketState {
    storyPoint: null | number;
    description: string;
    title: string;
}

export interface AppState {
    routing: RouterState;
    ticketList: TicketListState;
    addTicket: AddTicketState;
    loading: LoadingState;
    closeSprintDialog: CloseSprintDialogState;
}

const reduxStore: Store<AppState, AnyAction> = createStore<AppState, AnyAction, void, void>(appReducer, {}, composeWithDevTools<any, {}>(applyMiddleware(...middleware)));

export { reduxStore };
