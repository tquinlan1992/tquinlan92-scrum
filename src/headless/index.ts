import { reduxStore } from './store';

export { reduxStore };
