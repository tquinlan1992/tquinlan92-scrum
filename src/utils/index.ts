import * as _ from 'lodash';
import { connect, MapStateToPropsParam, MergeProps } from 'react-redux';
import { AppState } from '@headless/store';
import { ThunkDispatch } from 'redux-thunk';
import { AnyAction } from 'redux';

type PropsMap<M, K extends keyof M> = {[P in K]: M[P]};

export function pick<M extends object, K extends keyof M>(obj: M, ...props: K[]): PropsMap<M, K> {
    // @ts-ignore: TS2352: Type 'PartialDeep<M>' cannot be converted to type 'PropsMap<M, K>'.
    return _.pick<M>(obj, props) as PropsMap<M, K>;
}

export type Diff<T extends string, U extends string> = ({ [P in T]: P } & { [P in U]: never } & { [x: string]: never })[T];

export type AppStateThunkDispatch = ThunkDispatch<AppState, undefined, AnyAction>;
