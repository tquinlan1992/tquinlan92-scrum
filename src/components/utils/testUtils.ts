import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { AnyAction, isType, ActionCreator } from 'typescript-fsa';
import { AppState } from '../../headless/store';

export function getAnyJestFn() {
    return jest.fn() as any;
}

interface AnyActionPayload extends AnyAction {
    payload: any;
}

export function expectActionWithPayload(actionToTest: AnyActionPayload, expectedActionCreator: ActionCreator<any>, expectedPayload?: any) {
    expect(isType(actionToTest, expectedActionCreator)).toBe(true);
    expect(actionToTest.payload).toEqual(expectedPayload);
}


const middlewares = [thunk];
export const mockStore = configureStore(middlewares);

export function getMockStore(state: Partial<AppState>) {
    return mockStore(state);
}

export function expectCalledOnceWith(mockFunction: jest.Mock<{}>, calledWith?: any) {
    expect(mockFunction).toHaveBeenCalledTimes(1);
    const expectedParams = calledWith ? [calledWith] : [];
    expect(mockFunction.mock.calls[0]).toEqual(expectedParams);
}

export function mockClearAll(mockFunctions: jest.Mock<{}>[]) {
    mockFunctions.forEach(mockFunction => {
        mockFunction.mockClear();
    })
}