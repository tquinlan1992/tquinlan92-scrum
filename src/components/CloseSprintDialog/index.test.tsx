import { CloseSprintDialog } from './';
import { shallow } from 'enzyme';
import * as React from 'react';
import { getAnyJestFn } from '../utils/testUtils';
import { mapStateToProps } from './';
import { AppState } from '../../headless/store';

describe('test snapshot of component', () => {
    it('should render', () => {
        const props = {
            onRequestClose: jest.fn(),
            open: true,
            onSubmit: jest.fn(),
            sprintName: 'sprintName',
            closeSprint: getAnyJestFn(),
            setCloseSprintDialogState: getAnyJestFn(),
            setAddTicketState: getAnyJestFn(),
            reset: getAnyJestFn()
        };
        const result = shallow(<CloseSprintDialog {...props} />);
        expect(result).toMatchSnapshot();
    });
});

describe('mapStateToProps', () => {
    const state: Partial<AppState> = ({
        closeSprintDialog: {
            sprintName: 'sprintName'
        }
    });
    const ownProps = {
        onRequestClose: jest.fn(),
        open: false,
        onSubmit: jest.fn()
    }
    const props = mapStateToProps(state, ownProps)
    it('should return valid props', () => {
        expect(props).toEqual({
            sprintName: 'sprintName',
            ...ownProps
        })
    })
})
