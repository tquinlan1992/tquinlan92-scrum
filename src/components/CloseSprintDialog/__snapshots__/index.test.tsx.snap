// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test snapshot of component should render 1`] = `
<WithStyles(Dialog)
  fullWidth={true}
  onExit={[Function]}
  open={true}
  title="Dialog With Actions"
>
  <WithStyles(DialogTitle)
    id="form-dialog-title"
  >
    Close Sprint
  </WithStyles(DialogTitle)>
  <WithStyles(DialogContent)>
    <form
      autoComplete="off"
    >
      <WithStyles(FormControl)>
        <TextField
          autoFocus={true}
          fullWidth={true}
          id="name"
          label="Sprint Name"
          margin="dense"
          onChange={[Function]}
          required={true}
          select={false}
          type="text"
          value="sprintName"
          variant="standard"
        />
      </WithStyles(FormControl)>
    </form>
  </WithStyles(DialogContent)>
  <WithStyles(DialogActions)>
    <WithStyles(Button)
      color="primary"
      onClick={[Function]}
    >
      Cancel
    </WithStyles(Button)>
    <WithStyles(Button)
      color="primary"
      onClick={[Function]}
    >
      Close Sprint
    </WithStyles(Button)>
  </WithStyles(DialogActions)>
</WithStyles(Dialog)>
`;
